core-bundle/src/Twig/Studio/Operation/AbstractOperation.php

```diff
@@ -111,6 +113,14 @@ protected function getName(): string
         return $this->name;
     }
 
+    protected function invalidateTemplateCache(OperationContext $context): void
+    {
+        $this->container
+            ->get('contao.twig.studio.cache_invalidator')
+            ->invalidateCache($context->getIdentifier(), $context->getThemeSlug())
```

Comment:
```suggestion
            ?->invalidateCache($context->getIdentifier(), $context->getThemeSlug())
```

The service is optional, so theoretically it could be `null` here, right?

Reply:
`$this->container->get()` never returns `null`, instead it would throw `Psr\Container\NotFoundExceptionInterface`. If the container is an instance of `Symfony\Component\DependencyInjection\ContainerInterface` you can do `$this->container->get('contao.twig.studio.cache_invalidator', ContainerInterface::NULL_ON_INVALID_REFERENCE)` (then it would be null).

Reply:
Then why is the service added like this? `$services['contao.twig.studio.cache_invalidator'] = '?'.CacheInvalidator::class;`

Reply:
What I meant was: if you want to check for an optional service here, you either have to provide `ContainerInterface::NULL_ON_INVALID_REFERENCE` - or catch the exception instead. Adding `?` to `get()` alone won't work.

And as to _why_ these services are made "optional": we decided to do that in all our abstract controllers (https://github.com/contao/contao/pull/8424). I am assuming @m-vo is following the same reasoning here for the abstract operation.

Reply:
> Then why is the service added like this? $services['contao.twig.studio.cache_invalidator'] = '?'.CacheInvalidator::class;

Exactly, I did not even add this in the initial version. :slightly_smiling_face:  This was done in a PR where it was changed for every abstract controller. Each service in this class is _not_ optional and I would prefer to have the `'?'.` bit removed.

Reply:
Ah, yeah, I remember the discussion. Let's just keep the `?` then, so we don't have to discuss again 😄